File: 2020/06/main_test.go
package main

import (
	"testing"
)

func TestShouldCountUniqueAnswers(t *testing.T) {
	test := []struct {
		Answers []string
		Expect  int
	}{
		{
			Answers: []string{"abc"},
			Expect:  3,
		},
		{
			Answers: []string{"a", "b", "c"},
			Expect:  3,
		},
		{
			Answers: []string{"ab", "ac"},
			Expect:  3,
		},
		{
			Answers: []string{"a", "a", "a", "a"},
			Expect:  1,
		},
		{
			Answers: []string{"b"},
			Expect:  1,
		},
	}

	for i, in := range test {
		result := countUniqueAnswers(in.Answers)
		if result != in.Expect {
			t.Errorf("%d: got %d, expected %d", i, result, in.Expect)
		}
	}
}

func TestShouldCountAllSharedAnswers(t *testing.T) {
	test := []struct {
		Answers []string
		Expect  int
	}{
		{
			Answers: []string{"abc"},
			Expect:  3,
		},
		{
			Answers: []string{"a", "b", "c"},
			Expect:  0,
		},
		{
			Answers: []string{"ab", "ac"},
			Expect:  1,
		},
		{
			Answers: []string{"a", "a", "a", "a"},
			Expect:  1,
		},
		{
			Answers: []string{"b"},
			Expect:  1,
		},
	}

	for i, in := range test {
		result := countAllSharedAnswers(in.Answers)
		if result != in.Expect {
			t.Errorf("%d: got %d, expected %d", i, result, in.Expect)
		}
	}
}

func TestTask1ShouldFindResult(t *testing.T) {
	input := [][]string{
		{"abc"},
		{"a", "b", "c"},
		{"ab", "ac"},
		{"a", "a", "a", "a"},
		{"b"},
	}
	expected := 11

	result := Task1(input)
	if result != expected {
		t.Errorf("got %d, expected %d", result, expected)
	}
}

func TestTask2ShouldFindResult(t *testing.T) {
	input := [][]string{
		{"abc"},
		{"a", "b", "c"},
		{"ab", "ac"},
		{"a", "a", "a", "a"},
		{"b"},
	}
	expected := 6

	result := Task2(input)
	if result != expected {
		t.Errorf("got %d, expected %d", result, expected)
	}
}

File: 2021/05/src/main_test.zig
const std = @import("std");

const app = @import("./main.zig");

test "expect task 1 to yield correct result" {
    var input = [_]app.Vent{
        .{ .a = .{ .x = 0, .y = 9 }, .b = .{ .x = 5, .y = 9 } },
        .{ .a = .{ .x = 8, .y = 0 }, .b = .{ .x = 0, .y = 8 } },
        .{ .a = .{ .x = 9, .y = 4 }, .b = .{ .x = 3, .y = 4 } },
        .{ .a = .{ .x = 2, .y = 2 }, .b = .{ .x = 2, .y = 1 } },
        .{ .a = .{ .x = 7, .y = 0 }, .b = .{ .x = 7, .y = 4 } },
        .{ .a = .{ .x = 6, .y = 4 }, .b = .{ .x = 2, .y = 0 } },
        .{ .a = .{ .x = 0, .y = 9 }, .b = .{ .x = 2, .y = 9 } },
        .{ .a = .{ .x = 3, .y = 4 }, .b = .{ .x = 1, .y = 4 } },
        .{ .a = .{ .x = 0, .y = 0 }, .b = .{ .x = 8, .y = 8 } },
        .{ .a = .{ .x = 5, .y = 5 }, .b = .{ .x = 8, .y = 2 } },
    };
    const expected: u32 = 5;

    try std.testing.expect((try app.task1(std.testing.allocator, .{ .x = 10, .y = 10 }, input[0..])) == expected);
}

test "expect task 2 to yield correct result" {
    var input = [_]app.Vent{
        .{ .a = .{ .x = 0, .y = 9 }, .b = .{ .x = 5, .y = 9 } },
        .{ .a = .{ .x = 8, .y = 0 }, .b = .{ .x = 0, .y = 8 } },
        .{ .a = .{ .x = 9, .y = 4 }, .b = .{ .x = 3, .y = 4 } },
        .{ .a = .{ .x = 2, .y = 2 }, .b = .{ .x = 2, .y = 1 } },
        .{ .a = .{ .x = 7, .y = 0 }, .b = .{ .x = 7, .y = 4 } },
        .{ .a = .{ .x = 6, .y = 4 }, .b = .{ .x = 2, .y = 0 } },
        .{ .a = .{ .x = 0, .y = 9 }, .b = .{ .x = 2, .y = 9 } },
        .{ .a = .{ .x = 3, .y = 4 }, .b = .{ .x = 1, .y = 4 } },
        .{ .a = .{ .x = 0, .y = 0 }, .b = .{ .x = 8, .y = 8 } },
        .{ .a = .{ .x = 5, .y = 5 }, .b = .{ .x = 8, .y = 2 } },
    };
    const expected: u32 = 12;

    std.testing.log_level = .debug;

    try std.testing.expect((try app.task2(std.testing.allocator, .{ .x = 10, .y = 10 }, input[0..])) == expected);
}
